restricted_site_access.php

```diff
@@ -122,6 +122,32 @@ public static function add_actions() {
 
 		add_filter( 'pre_option_blog_public', array( __CLASS__, 'pre_option_blog_public' ), 10, 1 );
 		add_filter( 'pre_site_option_blog_public', array( __CLASS__, 'pre_option_blog_public' ), 10, 1 );
+		add_filter( 'application_password_is_api_request', array( __CLASS__, 'is_api_request' ) );
+	}
+
+	/**
+	 * Add filter to override application_password_is_api_request filter. This has to be done since rest API isn't been loaded until we call restrict_access.
+	 *
+	 * @param bool $original_value Original value passed by filter.
+	 * @return bool
+	 */
+	public static function is_api_request( $original_value ) {
+
+		$protocol = ( isset( $_SERVER['HTTPS'] ) && 'on' === $_SERVER['HTTPS'] ? 'https' : 'http' );
+		$request  = isset( $_SERVER['HTTP_HOST'] ) && sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) )
+		. isset( $_SERVER['REQUEST_URI'] ) && sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) );
+
+		// We need to determine current URL via php because $wp->request is not set at this point.
+		$current_request = esc_url_raw( "$protocol://$request" );
+
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( ! get_option( 'permalink_structure' ) && isset( $_GET['rest_route'] ) ) {
+			return true;
+		} elseif ( strpos( $current_request, home_url( '/wp-json' ) ) === 0 ) {
```

Comment:
This is my fault sorry, if pretty permalinks are `/post-name` we need to protect against someone writing a post with the slug `/wp-json-best-practices` and allowing the visitor to use application passwords.

As the base rest response works with both `/wp-json` and `/wp-json/` I've allowed for each.

It's the second condition that required the query string be removed in the suggestion above.

```suggestion
		} elseif (
			strpos( $current_request, home_url( '/wp-json/' ) ) === 0
			|| home_url( '/wp-json' ) === $current_request
		) {
```